clang/test/CodeGen/debug-dir-win-pe-ltcg-string.c

```diff
@@ -0,0 +1,13 @@
+// This test checks if Window PE file compiled with -flto option contains a magic 
+// string "LTCG" to indicate LTO compilation.
+
+// REQUIRES: system-windows
+
+// RUN: %clang --target=x86_64-pc-windows-msvc -flto -fuse-ld=lld %s -o %t.exe
```

Comment:
Is there anything that is x64 specific? Can we support this in ARM64?

Reply:
Good point, this should also work on ARM

Reply:
> Note that this test as such doesn't fit the testing separation within LLVM.
> 
> As it stands, this test does compilation with Clang, linking with LLD, and inspects the results.
> 
> This is generally not allowed in Clang tests. A Clang test normally should only test that producing LLVM IR works correctly. An LLVM level test can test that IR -> object files does the right thing. And an LLD test can test that object file -> executable does the right thing.
> 
> But there's in general no guarantee that e.g. LLD is available while testing Clang, or that you have the right headers/libraries that might be expected while compiling. (In this case, the generated object files can end up requiring `msvcrt.lib` or `libcmt.lib` or similar libraries that aren't available in general, when cross compiling.)

My intention was to create a holistic test that tests the feature from clang invocation to checking that magic strings are emitted. I already have a lld tests; I will change the pgi/pgu to be solely a clang test, checking for .pgi/.pgu section existence. For LTO case, I need to do this this way, since I have to test that lld emits "LTCG" when given an LTOed object file. Do you have any suggestions where could I move this test? Would moving the "LTCG" test to lld be OK?

Reply:
Yes, I understand that you’d want to test the full chain - that would be highly valuable, but that’s unfortunately generally much less portable, and doesn’t work as all these tools are in separate subprojects that are independent.

In clang you can have a test that runs clang on a C file and produces LLVM IR - you can check that the produced IR contains the relevant key elements. A LLVM test, if necessary (but I guess it isn’t necessary here), can compile the IR to an object file, and you can check that it contains the necessary bits. And in LLD, you can have the LTO test - you’d have LLVM IR that gets built into a bitcode object file (i.e. LTO object file), which you can link and check the output binary properties.

That kind of holistic tests would need to go in a separate subproject for end-to-end tests - we do have that I think, but I’m not sure how much it is used.

Reply:
llvm-test-suite is the right place for that. And I think it should be in use, just not as much as the normal test suite.

Reply:
> Yes, I understand that you’d want to test the full chain - that would be highly valuable, but that’s unfortunately generally much less portable, and doesn’t work as all these tools are in separate subprojects that are independent.
> 
> In clang you can have a test that runs clang on a C file and produces LLVM IR - you can check that the produced IR contains the relevant key elements. A LLVM test, if necessary (but I guess it isn’t necessary here), can compile the IR to an object file, and you can check that it contains the necessary bits. And in LLD, you can have the LTO test - you’d have LLVM IR that gets built into a bitcode object file (i.e. LTO object file), which you can link and check the output binary properties.
> 
> That kind of holistic tests would need to go in a separate subproject for end-to-end tests - we do have that I think, but I’m not sure how much it is used.

I have changed the pgi test to emit object file and check for magic section. I will try to do something similar to the ltcg/pgu test, but I have to provide them with LTO object file and instrumentation file respectively.

Reply:
> llvm-test-suite is the right place for that

That’s not exactly what I meant - llvm-test-suite has full test applications where it’s more about executing the built code - while for the same style of build-and-inspect lit tests, there is llvm-project/cross-project-tests - which is rather new, and I’m not sure about how much it actually is executed. (As for llvm-test-suite, it’s not completely buildable on windows, as far as I know.)

Regarding an all-in-one compile-and-link test; one risk (which of course is possible to test and handle) is that the test implicitly requires you to have MSVC/WinSDK files available (if the compiler injects such defaultlib directives) - so even if the test has `REQUIRES: system-windows` or similar, it could break in an mingw environment. So while the isolated stepwise unit tests feel more brittle, they ensure that those tests can run in any environment where we can compile clang and lld, as we exactly control the inputs to each step.

I.e. I’m not sure that we really do need the end-to-end tests unless you feel very strongly about it - the stepwise isolated unit tests do provide the same coverage. Back when I was new to writing tests within this ecosystem, I also wanted to test things that way, but I’ve come to appreciate the elegance of the split up tests.

Reply:
> > llvm-test-suite is the right place for that
> 
> That’s not exactly what I meant - llvm-test-suite has full test applications where it’s more about executing the built code - while for the same style of build-and-inspect lit tests, there is llvm-project/cross-project-tests - which is rather new, and I’m not sure about how much it actually is executed. (As for llvm-test-suite, it’s not completely buildable on windows, as far as I know.)
> 
> Regarding an all-in-one compile-and-link test; one risk (which of course is possible to test and handle) is that the test implicitly requires you to have MSVC/WinSDK files available (if the compiler injects such defaultlib directives) - so even if the test has `REQUIRES: system-windows` or similar, it could break in an mingw environment. So while the isolated stepwise unit tests feel more brittle, they ensure that those tests can run in any environment where we can compile clang and lld, as we exactly control the inputs to each step.
> 
> I.e. I’m not sure that we really do need the end-to-end tests unless you feel very strongly about it - the stepwise isolated unit tests do provide the same coverage. Back when I was new to writing tests within this ecosystem, I also wanted to test things that way, but I’ve come to appreciate the elegance of the split up tests.

I see what you mean, I have separated the tests: the Codegen ones check if magic sections are emitted, and the lld ones check for magic strings from these sections (or from bitcode in case of LTO test)